sihl/src/contract_queue.ml

```diff
@@ -1,64 +1,192 @@
+      | Failed -> "failed"
+      | Cancelled -> "cancelled"
+    ;;
 
-  (** [register_jobs jobs] registers jobs that can be dispatched.
+    let of_string str =
+      match str with
+      | "pending" -> Ok Pending
+      | "succeeded" -> Ok Succeeded
+      | "failed" -> Ok Failed
+      | "cancelled" -> Ok Cancelled
+      | _ -> Error (Printf.sprintf "Unexpected job status %s found" str)
+    ;;
+  end
 
-      Only registered jobs can be dispatched. Dispatching a job that was not
-      registered does nothing. *)
-  val register_jobs : 'a t list -> unit Lwt.t
+  type t =
+    { id : string
+    ; name : string
+    ; input : string
+    ; tries : int
+    ; next_run_at : Ptime.t
+    ; max_tries : int
+    ; status : Status.t
+    ; last_error : string option
+    ; last_error_at : Ptime.t option
+    }
+  [@@deriving show]
 
-  val register : ?jobs:'a t list -> unit -> Core_container.Service.t
+  let create input delay now (job : job) =
+    let name = job.name in
+    let next_run_at =
+      match delay with
+      | Some delay -> Option.value (Ptime.add_span now delay) ~default:now
+      | None -> now
+    in
+    let max_tries = job.max_tries in
+    { id = Uuidm.v `V4 |> Uuidm.to_string
+    ; name
+    ; input
+    ; tries = 0
+    ; next_run_at
+    ; max_tries
+    ; status = Status.Pending
+    ; last_error = None
+    ; last_error_at = None
+    }
+  ;;
 
-  include Core_container.Service.Sig
+  let is_pending job_instance =
+    match job_instance.status with
+    | Status.Pending -> true
+    | _ -> false
+  ;;
+
+  let incr_tries job_instance =
+    { job_instance with tries = job_instance.tries + 1 }
+  ;;
+
+  let update_next_run_at retry_delay job_instance =
+    let next_run_at =
+      match Ptime.add_span job_instance.next_run_at retry_delay with
+      | Some date -> date
+      | None -> failwith "Can not determine next run date of job"
+    in
+    { job_instance with next_run_at }
+  ;;
+
+  let should_run job_instance now =
+    let tries = job_instance.tries in
+    let max_tries = job_instance.max_tries in
+    let next_run_at = job_instance.next_run_at in
+    let has_tries_left = tries < max_tries in
+    let is_after_delay = not (Ptime.is_later next_run_at ~than:now) in
+    let is_pending = is_pending job_instance in
+    is_pending && has_tries_left && is_after_delay
+  ;;
 end
 
-(* Common *)
-
-let to_sexp { name; max_tries; retry_delay; _ } =
-  let open Sexplib0.Sexp_conv in
-  let open Sexplib0.Sexp in
-  List
-    [ List [ Atom "name"; sexp_of_string name ]
-    ; List [ Atom "max_tries"; sexp_of_int max_tries ]
-    ; List
-        [ Atom "retry_delay"
-        ; sexp_of_string (Core_time.show_duration retry_delay)
-        ]
-    ]
+let default_tries = 5
+let default_retry_delay = Core_time.Span.minutes 1
+
+let default_error_handler id name input msg =
+  Lwt.return
+  @@ Logs.err (fun m ->
+         m
+           "Job with id '%s' and name '%s failed for input '%s': %s"
```

Comment:
Add `'` after `name '%s` and put the last `%s` in `''` too please

Reply:
Fixed.

Reply:
Did you not put the last `%s` in `''` on purpose?

Reply:
Yes because there is a `:` before